File: day_02/go/day_02.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
)

func loadFile(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	serials := make([]string, 0)

	reader := bufio.NewReader(file)
	for {
		line, _, err := reader.ReadLine()
		if err != nil {
			if err == io.EOF {
				return serials
			}
			log.Fatal(err)
		}
		trimmedLine := strings.TrimSpace(string(line))
		serials = append(serials, trimmedLine)
	}
}

func countPairsAndTriplets(s string) (int, int) {
	count := map[rune]int{}
	runes := []rune(s)
	for _, r := range runes {
		if v, ok := count[r]; ok {
			count[r] = v + 1
		} else {
			count[r] = 1
		}
	}

	pairs := 0
	triplets := 0
	for _, v := range count {
		switch v {
		case 2:
			pairs++
		case 3:
			triplets++
		}
	}
	return pairs, triplets
}

func part1(serials []string) {
	pairs := 0
	triplets := 0
	for _, serial := range serials {
		p, t := countPairsAndTriplets(serial)
		if p >= 1 {
			pairs++
		}
		if t >= 1 {
			triplets++
		}
	}
	checksum := pairs * triplets
	fmt.Printf("Part1: %v * %v = %v\n", pairs, triplets, checksum)
}

func numberOfNonEqualCharacters(s1 string, s2 string) (int, string) {
	r1 := []rune(s1)
	r2 := []rune(s2)
	count := 0
	rest := make([]rune, 0)
	for i := 0; i < len(r1); i++ {
		if r1[i] != r2[i] {
			count++
		} else {
			rest = append(rest, r1[i])
		}
	}
	return count, string(rest)
}

func part2(serials []string) {
	for i, serial1 := range serials {
		for _, serial2 := range serials[i:] {
			if serial1 == serial2 {
				continue
			}
			count, rest := numberOfNonEqualCharacters(serial1, serial2)
			if count == 1 {
				fmt.Println(count, rest)
			}
		}
	}
}

func main() {
	serials := loadFile("input.txt")
	part1(serials)
	part2(serials)
}

File: day_02/scala/src/main/scala/Day02.scala
import scala.io.Source

object Day02 {
  def loadFile(filename: String): Seq[String] =
    Source
      .fromFile(filename)
      .getLines()
      .map(_.trim)
      .toSeq

  def countPairsAndTriplets(str: String): (Int, Int) = {
    val charCount = str.foldLeft(Map.empty[Char, Int]) { (state, value) =>
      val currentCount = state.getOrElse(value, 0)
      state + (value -> (currentCount + 1))
    }
    charCount.values.foldLeft((0, 0)) { (state, value) =>
      value match {
        case 2 => (state._1 + 1, state._2)
        case 3 => (state._1, state._2 + 1)
        case _ => state
      }
    }
  }

  def part1(serials: Seq[String]): Unit = {
    val (pairs, triplets) = serials.foldLeft((0, 0)) { (count, serial) =>
      countPairsAndTriplets(serial) match {
        case (p, t) if p >= 1 && t >= 1 => (count._1 + 1, count._2 + 1)
        case (p, _) if p >= 1           => (count._1 + 1, count._2)
        case (_, t) if t >= 1           => (count._1, count._2 + 1)
        case _ => count
      }
    }
    val checksum = pairs * triplets
    println(s"Part1: $pairs * $triplets = $checksum")
  }

  def numberOfNonEqualCharacters(s1: String, s2: String): (Int, String) = {
    val (count, rest) = s1.zip(s2).foldLeft((0, List.empty[Char])) { case (state, (c1, c2)) =>
        if (c1 != c2) {
          (state._1 + 1, state._2)
        } else {
          (state._1, c1 +: state._2)
        }
    }
    (count, rest.reverse.mkString)
  }

  def part2(serials: Seq[String]): Unit = {
    serials.zipWithIndex.foreach { case (serial1, i) =>
      serials.drop(i).foreach { serial2 =>
        val (count, rest) = numberOfNonEqualCharacters(serial1, serial2)
        if (count == 1) {
          println(s"$count : $rest")
        }
      }
    }
  }

  def main(args: Array[String]): Unit = {
    val serials = loadFile("../input.txt")
    part1(serials)
    part2(serials)
  }
}
